### PR TITLE
Stop planner if the goal is cancelled

### PR DESCRIPTION
**Pull request by bektaskemal**
---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | [4139](https://github.com/ros-planning/navigation2/issues/4139) |
| Primary OS tested on | Ubuntu|
| Robotic platform tested on | Gazebo TB3 simulation |
| Does this PR contain AI generated software? | No |

---

## Description of contribution in a few bullet points
* Stops the planner execution if the goal is cancelled

## Description of documentation updates required from your changes

---

#### For Maintainers: 
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

### CODE REVIEW

File: nav2_smac_planner/src/a_star.cpp [hunk 1]
```diff
@@ -294,6 +295,11 @@ bool AStarAlgorithm<NodeT>::createPath(
       }
     }
 
+    // Check for cancel every Nth iteration
+    if (iterations % _cancel_check_interval == 0 && cancel_checker()) {
```
Comment (SteveMacenski) [inline nav2_smac_planner/src/a_star.cpp:299]: Is 500 vs 5000 a big enough difference to not have it in the same timing interval branch (and possibly rename to be more generic)?

File: nav2_navfn_planner/src/navfn.cpp [hunk 2]
```diff
@@ -667,6 +672,10 @@ NavFn::propNavFnAstar(int cycles)
 
   // do main cycle
   for (; cycle < cycles; cycle++) {  // go for this many cycles, unless interrupted
+    if (cycle % cancel_check_interval == 0 && cancelChecker()) {
+      throw nav2_core::PlannerCancelled("Planner was cancelled");
```
Comment (SteveMacenski) [inline nav2_navfn_planner/src/navfn.cpp:676]: I don't see where this exception is caught specially to handle this condition now that its created

Comment (SteveMacenski) [reply nav2_navfn_planner/src/navfn.cpp:676]: It should be in the catch list and if caught, do what we do in the cancel function: https://github.com/ros-planning/navigation2/blob/9e0e1f6b785d89cebd876dfbe1fbed7bea8c644f/nav2_planner/src/planner_server.cpp#L298

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Overall, LGTM!

File: nav2_navfn_planner/src/navfn.cpp [hunk 3]
```diff
@@ -667,6 +672,10 @@ NavFn::propNavFnAstar(int cycles)
 
   // do main cycle
   for (; cycle < cycles; cycle++) {  // go for this many cycles, unless interrupted
+    if (cycle % cancel_check_interval == 0 && cancelChecker()) {
```
Comment (SteveMacenski) [inline nav2_navfn_planner/src/navfn.cpp:675]: Where is cancel_check_interval set and parameterized?

File: nav2_theta_star_planner/src/theta_star.cpp [hunk 4]
```diff
@@ -60,6 +61,11 @@ bool ThetaStar::generatePath(std::vector<coordsW> & raw_path)
   while (!queue_.empty()) {
     nodes_opened++;
 
+    if (nodes_opened % cancel_check_interval_ && cancel_checker()) {
```
Comment (SteveMacenski) [inline nav2_theta_star_planner/src/theta_star.cpp:64]:
```suggestion
    if (nodes_opened % cancel_check_interval_ == 0 && cancel_checker()) {
```

File: nav2_navfn_planner/src/navfn.cpp (hunk 3 above)
Comment (bektaskemal) [reply nav2_navfn_planner/src/navfn.cpp]: I parametrized it now for other planners but it wasn't so clear how nicely do it for NavFn, so I made it constant for now.

File: nav2_navfn_planner/src/navfn.cpp (hunk 2 above)
Comment (bektaskemal) [reply nav2_navfn_planner/src/navfn.cpp:676]: Done.

File: nav2_smac_planner/src/a_star.cpp (hunk 1 above)
Comment (bektaskemal) [reply nav2_smac_planner/src/a_star.cpp]: I am not very sure about numbers so I am open to suggestions for default values, but I made it a parameter now so seperate branch is needed.

File: nav2_navfn_planner/src/navfn.cpp (hunk 3 above)
Comment (SteveMacenski) [reply nav2_navfn_planner/src/navfn.cpp]: Where is that constant set? As far as I see, its just instantiated (so 0) but never set

File: nav2_smac_planner/src/a_star.cpp (hunk 1 above)
Comment (SteveMacenski) [reply nav2_smac_planner/src/a_star.cpp]: Lets not then, use the same parameter branch to check for all potential preemption conditions like cancel / timeout

File: nav2_smac_planner/src/smac_planner_lattice.cpp [hunk 5]
```diff
@@ -504,15 +509,18 @@ SmacPlannerLattice::dynamicParametersCallback(std::vector<rclcpp::Parameter> par
...
-      }
-    } else if (name == _name + ".max_on_approach_iterations") {
-      reinit_a_star = true;
-      _max_on_approach_iterations = parameter.as_int();
-      if (_max_on_approach_iterations <= 0) {
-        RCLCPP_INFO(
-          _logger, "On approach iteration selected as <= 0, "
-          "disabling tolerance and on approach iterations.");
-        _max_on_approach_iterations = std::numeric_limits<int>::max();
+      } else if (name == _name + ".max_on_approach_iterations") {
```
Comment (SteveMacenski) [inline nav2_smac_planner/src/smac_planner_lattice.cpp:512]: Appears to be misformatting, I'm sure CI is complaining about linting

**Review by SteveMacenski — CHANGES_REQUESTED** [review]
Overall, looks very good thank you for this great contribution!

A couple of small things to finish it up
- Documentation updates for new parameter for all appropriate algorithms. Add to Readmes (if they have a param list) & navigation.ros.org's configuration guide entries for the algorithms to add their params
- Migration guide entry on navigation.ros.org on Planner plugin API + new param and how to use 
- I'm noticing alot of code cov warnings on lack of test coverage on this PRs diffs. We need things to be 90% tested, it seems like the dynamic parameter changes weren't added to their tests to exercise those lines, nor is cancel actually called on any of the algorithms as part of test to make sure that it does actually return early -- thus also that exception itself / planner server handling isn't exercised either

File: nav2_smac_planner/src/smac_planner_lattice.cpp (hunk 5 above)
Comment (bektaskemal) [reply nav2_smac_planner/src/smac_planner_lattice.cpp:512]: I believe there was a mistake here and this line wasn't inside PARAMETER_INTEGER branch, so I fixed it.

**Review by SteveMacenski — APPROVED** [review]
LGTM, just one set of important lines that aren't covered, the actual catching of the exception and checking that the correct state happens in the planner server. 

We have some tests that trigger the others in https://github.com/ros-planning/navigation2/tree/main/nav2_system_tests/src/error_codes. I think you can just add the new plugin for cancel and check that its terminated correctly as a cancel state

Sorry to be a stickler on this one - these are just important systems to make sure are tested and have near full coverage in case of regression.

**Comment by bektaskemal** [comment]
> We have some tests that trigger the others in https://github.com/ros-planning/navigation2/tree/main/nav2_system_tests/src/error_codes. I think you can just add the new plugin for cancel and check that its terminated correctly as a cancel state

If planner returns PlannerCancelled without a cancel request, the result status was ABORTED since we call terminate_all. So I tested it with an actual cancel request. 
https://github.com/ros-planning/navigation2/pull/4148/files#diff-7d5b834e9aba5d3d110b02c975dc4c997b854ce7e92bdeb03d68108fa3071df1R120-R124

**Review by SteveMacenski — APPROVED** [review]

**Comment by SteveMacenski** [comment]
Great - a couple of tests failed - I don't think they're related to you - but rerunning CI just to check. I expect to be able to merge this before the end of today

**Pull request merged**